sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SharedAccessSignatureBuilder.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Azure.Iot.Hub.Service.Authentication
+{
+    /// <summary>
+    /// Builds the shared access signature based on the access policy passed.
+    /// </summary>
+    internal sealed class SharedAccessSignatureBuilder
+    {
+        public SharedAccessSignatureBuilder()
+        {
+            TimeToLive = TimeSpan.FromMinutes(5);
+        }
+
+        public string KeyName { get; set; }
+
+        public string Key { get; set; }
+
+        public string Target { get; set; }
+
+        public TimeSpan TimeToLive { get; set; }
+
+        public string ToSignature()
+        {
+            return BuildSignature(KeyName, Key, Target, TimeToLive);
+        }
+
+        private static string BuildSignature(string keyName, string key, string target, TimeSpan timeToLive)
+        {
+            string expiresOn = BuildExpiresOn(timeToLive);
+            string audience = WebUtility.UrlEncode(target);
+            List<string> fields = new List<string>
+            {
+                audience,
+                expiresOn
+            };
+
+            // Example string to be signed:
+            // dh://myiothub.azure-devices.net/a/b/c?myvalue1=a
+            // <Value for ExpiresOn>
+
+            string signature = Sign(string.Join("\n", fields), key);
+
+            // Example returned string:
+            // SharedAccessSignature sr=ENCODED(dh://myiothub.azure-devices.net/a/b/c?myvalue1=a)&sig=<Signature>&se=<ExpiresOnValue>[&skn=<KeyName>]
+
+            var buffer = new StringBuilder();
+            buffer.AppendFormat(CultureInfo.InvariantCulture, "{0} {1}={2}&{3}={4}&{5}={6}",
+                SharedAccessSignatureConstants.SharedAccessSignature,
+                SharedAccessSignatureConstants.AudienceFieldName, audience,
+                SharedAccessSignatureConstants.SignatureFieldName, WebUtility.UrlEncode(signature),
+                SharedAccessSignatureConstants.ExpiryFieldName, WebUtility.UrlEncode(expiresOn));
+
+            if (!string.IsNullOrWhiteSpace(keyName))
+            {
+                buffer.AppendFormat(CultureInfo.InvariantCulture, "&{0}={1}",
```

Comment:
Out of curiosity, why this approach instead of:

```csharp
buffer.Append($"&{SharedAccessSignatureConstants.KeyNameFieldName}={WebUtility.UrlEncode(keyName)}");
```

Reply:
This is definitely better, I reused the code from elsewhere and didn't think to change.